daemon/cluster/cluster.go

```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 	return services, nil
 }
 
+// imageWithDigestString takes an image such as name or name:tag
+// and returns the image pinned to a digest, such as name@sha256:34234...
+func (c *Cluster) imageWithDigestString(ctx context.Context, image, encodedAuth string) (string, error) {
+	// TODO(nishanttotla): create AuthConfig from encodedAuth
+
+	ref, err := reference.ParseNamed(image)
+	if err != nil {
+		logrus.Warnf("unable to parse image name: %s", err.Error())
+		return "", err
+	}
+	// only pin if not a canonical reference (i.e. with digest)
+	if _, ok := ref.(reference.Canonical); !ok {
+		// add default tag if not present
+		if reference.IsNameOnly(ref) {
+			ref = reference.WithDefaultTag(ref)
+		}
+		namedTaggedRef, _ := ref.(reference.NamedTagged)
```

Comment:
namedTaggedRef, ok := ref.(reference.NamedTagged)

Return an error if `ok` is false (should not be possible currently, but the reference package could change in the future).